src/cpp/rtps/RTPSDomain.cpp

```diff
@@ -212,22 +212,22 @@ RTPSParticipant* RTPSDomain::createParticipant(
 bool RTPSDomain::removeRTPSParticipant(
         RTPSParticipant* p)
 {
-    if (p != nullptr)
+    for (auto it = m_RTPSParticipants.begin(); it != m_RTPSParticipants.end(); ++it)
     {
-        p->mp_impl->disable();
-
-        std::unique_lock<std::mutex> lock(m_mutex);
-        for (auto it = m_RTPSParticipants.begin(); it != m_RTPSParticipants.end(); ++it)
+        if (it->first == p)
         {
-            if (it->second->getGuid().guidPrefix == p->getGuid().guidPrefix)
+            if (p->mp_impl != nullptr)
             {
-                RTPSDomain::t_p_RTPSParticipant participant = *it;
-                m_RTPSParticipants.erase(it);
-                m_RTPSParticipantIDs.erase(m_RTPSParticipantIDs.find(participant.second->getRTPSParticipantID()));
-                lock.unlock();
-                removeRTPSParticipant_nts(participant);
-                return true;
+                p->mp_impl->disable();
             }
+
+            std::unique_lock<std::mutex> lock(m_mutex);
```

Comment:
This lock should be taken before traversing the list of participants, otherwise it may be changed during the processing, invalidating the iterators. However, this will have the effect of calling `ChannelResource::disable` with the mutex locked (during the `disable`). I don't think this is any problem, but be aware of it.